src/sage/schemes/projective/coherent_sheaf.py

```diff
@@ -0,0 +1,323 @@
+
+    def cohomology(self, r=0):
+        """
+        Return the dimension of the `r`-th cohomology as a vector space.
+
+        INPUT:
+
+        - ``r`` -- (default: 0) a non-negative integer
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: sheaf = X.structure_sheaf()
+            sage: sheaf.cohomology(0)
+            1
+            sage: sheaf.cohomology(1)
+            3
+            sage: sheaf.cohomology(2)
+            0
+        """
+        return self._cohomology.h(r)
+
+    def twist(self, t=0):
+        r"""
+        Return the twisted sheaf `F(n)` of this sheaf `F`.
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: sh = X.structure_sheaf()
+            sage: sh.twist(1).euler_characteristic()
+            2
+            sage: sh.twist(2).euler_characteristic()
+            6
+        """
+        return type(self)(self._base_scheme, self._module, self._twist + t)
+
+    def euler_characteristic(self):
+        """
+        Return the Euler characteristic of this coherent sheaf.
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: X = P2.subscheme(x^4 + y^4 + z^4)
+            sage: sh = X.structure_sheaf()
+            sage: sh.euler_characteristic()
+            -2
+        """
+        d = self._base_scheme.dimension()
+        chi = 0
+        for r in range(d + 1):  # for Grothendieck's vanishing theorem
+            d = self.cohomology(r)
+            if r % 2:
+                chi = chi - d
+            else:
+                chi = chi + d
+        return chi
+
+
+class CoherentSheaf_on_projective_space(CoherentSheaf):
+    r"""
+    Coherent sheaf on a projective space.
+
+    EXAMPLES::
+
+        sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+        sage: P2.structure_sheaf()
+        Coherent sheaf on Projective Space of dimension 2 over Rational Field
+    """
+    @cached_property
+    def _cohomology(self):
+        """
+        This property keeps the cohomology object for this sheaf.
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+            sage: P2.structure_sheaf()._cohomology
+            Maruyama Complex induced from S(0) <-- 0
+        """
+        from sage.schemes.projective.cohomology import MaruyamaComplex
+        return MaruyamaComplex(self._module, twist=self._twist)
+
+
+class CoherentSheaf_on_projective_subscheme(CoherentSheaf):
+    r"""
+    Coherent sheaf on a projective subscheme.
+
+    EXAMPLES::
+
+        sage: P2.<x,y,z> = ProjectiveSpace(QQ, 2)
+        sage: X = P2.subscheme(x^4 + y^4 + z^4)
+        sage: X.structure_sheaf()
+        Coherent sheaf on Closed subscheme of Projective Space of dimension 2
+         over Rational Field defined by: x^4 + y^4 + z^4
+    """
+    @cached_property
```

Comment:
Same comment as above. Although because of the `image_to_ambient_space`, this would make more sense to do lazily here.

Reply:
I don't understand. `MaruyamaComplex` is constructed lazily, when requested via `cohomology()`, and then cached.

Reply:
It has to do with the call to `image_to_ambient_space`, which would be an argument for keeping the lazily construction in this class (rather than in `MaruyamaComplex`).

Reply:
The attribute `_cohomology` of `CoherentSheaf_xxx` class has `@cached_property`. 

So the object of class `MaruyamaMethod` is constructed when requested (lazily), and it is kept both in  the attribute `_cohomology` of `CoherentSheaf_on_projective_subscheme` and the attribute `_cohomology` of `CoherentSheaf_on_projective_space`. I see no problem here.

> It is better design to have `MaruyamaMethod` be a `UniqueRepresentation`.

Sorry. I still don't get it. 

`UniqueRepresentation` is used when returning the same object for the repeated requests with the same input is desirable.  I see no situation like that here.